### PR TITLE
dts: sensor: remove lis3dh binding

### LINKED ISSUE #31253: lis3dh driver support is confusing
**Describe the bug**
The ST lis3dh sensor device does appear to be supported but using the lis2dh driver, however the way it's been implemented is incredibly confusing and has wasted about an hour of my time this morning trying to figure it out, I'm not the only person that finds this confusing as noted in the original change https://github.com/zephyrproject-rtos/zephyr/pull/12373

The issue is that the lis2dh driver can be used for multiple devices, but there is a lis3dh driver binding which falls flat with the following build message:

    zephyr/drivers/sensor/lis2dh/lis2dh.c:353:2: warning: #warning "LIS2DH driver enabled without any devices" [-Wcpp]
      353 | #warning "LIS2DH driver enabled without any devices"
          |  ^~~~~~~

To get it to work, you must change to the lis2dh binding, my personal opinion is that this is annoying and bad, the lis3dh binding should be enough for the driver to work with it.
I haven't tested the other boards but I can see 96b_wistrio and rak5010_nrf52840 both only specify lis3dh bindings so I assume those actually do not work either.

**To Reproduce**
1. Configure a board with an lis3dh using:

    lis3dh@18 {
    	compatible = "st,lis3dh";
    	label = "LIS3DH";
    	reg = <0x18>;
    };

2. Configure any project to use board and LIS2DH sensor
3. Build, it will output an error about no lis2dh sensors being present
4. Change DTS line:
    	compatible = "st,lis2dh";
5. Rebuild, error vanishes and device communication works

**Expected behavior**
The sensor to work out the box with the lis3dh binding

**Impact**
Annoyance

**Environment (please complete the following information):**
 - OS: Windows 7
 - Toolchain: GNU GCC from ARM (version 9.3.1, 2020 Q2 update)
 - Commit SHA or Version used: f4a26837c8cb6d39f3307193068c3163b20ac584

### PR DESCRIPTION
**Pull request by pabigot**
While there once was a separate driver for st,lis3dh support for that variant was merged into lis2dh.  That driver only recognizes devices that are identified as st,lis2dh.  Remove the lis3dh binding, and update in-tree devicetree nodes to add the compatible required by the driver.

Also st,lis2dh12 same thing.  Also propose standard documentation block to relate compatibles with bindings and drivers.

(NB: Prior to this I verified that neither of the affected in-tree boards could build the lis2dh sample because the required devicetree binding macros were not generated.)

Fixes #31253.

### CODE REVIEW

**Review by galak — DISMISSED** [review]

**Review by galak — COMMENTED** [review]
Should we add a comment to the lis2dh binding/driver that it works for lis3dh?

**Comment by pabigot** [comment]
> Should we add a comment to the lis2dh binding/driver that it works for lis3dh?

Yes.  I've done that in a second commit, which @mbolivar-nordic should confirm.  This may be a path to finally resolving #19904.

I think there were other proposals to address that issue but I don't see them in tree; if they're in progress I can try to follow that standard instead.

Also noticed, since #19904 reminded me, that this also has to be done for lis2dh12, though the in-tree devicetree sources for that were correct.

**Review by MaureenHelm — DISMISSED** [review]
@avisconti please review

**Comment by avisconti** [comment]
@pabigot @mhelm 
To be honest I'm a little bit confused about the idea behind it. Let me try
to put some comments here to reconstruct the situation:

 1. The lis2dh driver is common between following devices:
        - LIS2DH
        - LIS3DH, 
        - LSM303DLHC   
        - LIS2DH12
        - LSM303AGR

      So, the current commit is only addressing a partial set of devices.

  2. Nevertheless, I remember an issue that was raised in the past by me, related to proper 
       documenting devices that are sharing same driver, but then for some reason I lost the
       track of the solution proposed.
       I remember that @erwango was proposing to generate documentation through bindings
       and @mbolivar-nordic did some implementations here: #30104.
       Now you are claiming that this PR is going into that direction, but it seems that it is removing        
       the binding instead. Can you put some light on it?
       
 Thanks and sorry if I lost some info in the meantime

**Comment by pabigot** [comment]
@avisconti Thanks for pointing to #30104; I couldn't find it.

I don't think the intent was ever to have a separate binding file for every compatible; that would be the wrong path.  All the way back to #19904 what we've really needed was to have multiple compatibles identified in the binding file.  So that's the path I'm following, hopefully it's not inconsistent with where #30104 ends up going.

I can add the other devices to this, I just wasn't aware of everything the current driver supports.  Is that list documented somewhere that I should have found?   Basically I want to be able to `git grep st,lsm303agr-accel` (for example) and find out how to describe it in devicetree (which binding and compatible must be present in the multi-element compatible list), and which driver supports it.

**Comment by pabigot** [comment]
I've updated this with some significant changes:
* Instead of removing the nodes from the test devicetree, retain them but correct their compatibles to also include LIS2DH
* Correct the compatible lines in various in-tree devicetree sources where the LIS2DH driver supplies the implementation
* Add documentation of the LSM303 variant accelerometer compatibles as falling under `st,lis2dh`

This should be reviewed as an alternative or adjunct to  #30104 per https://github.com/zephyrproject-rtos/zephyr/pull/31450#issuecomment-764787635

File: dts/bindings/sensor/st,lis2dh-i2c.yaml [hunk 1]
```diff
@@ -2,7 +2,30 @@
 # SPDX-License-Identifier: Apache-2.0
 
 description: |
-    STMicroelectronics LIS2DH 3-axis accelerometer accessed through I2C bus
+    STMicroelectronics LIS2DH 3-axis accelerometer accessed through I2C bus.
+
+    This binding is supported by CONFIG_LIS2DH.
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/st,lis2dh-i2c.yaml:7]: FYI, the reason I didn't take this approach is because the bindings are at least notionally OS agnostic.

As a practical matter, it seems unlikely to me that anybody besides Zephyr will ever be using this bindings language, but I don't know. @b0661 ?

**Comment by avisconti** [comment]
> I can add the other devices to this, I just wasn't aware of everything the current driver supports. Is that list documented somewhere that I should have found? Basically I want to be able to `git grep st,lsm303agr-accel` (for example) and find out how to describe it in devicetree (which binding and compatible must be present in the multi-element compatible list), and which driver supports it.

The list is inside $ZEPHYR_BASE/drivers/sensor/lis2dh/Kconfig, but is basically what I already said.

**Comment by avisconti** [comment]
> I've updated this with some significant changes:
> 
> * Instead of removing the nodes from the test devicetree, retain them but correct their compatibles to also include LIS2DH
> * Correct the compatible lines in various in-tree devicetree sources where the LIS2DH driver supplies the implementation
> * Add documentation of the LSM303 variant accelerometer compatibles as falling under `st,lis2dh`
> 
> This should be reviewed as an alternative or adjunct to #30104 per [#31450 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/31450#issuecomment-764787635)

First of all let me explain to you what is what I'm asking for at the first place. I was requested by my management to provide more visibility to lis2dh12, which is what our marketing is promoting. As a matter of fact lis2dh12 is supported in zephyr through lis2dh, yet there are users that are asking us for support. So, my original request was to create a proper documentation for such cases, where a single driver is supporting multiple devices. After that I left the discussion with the understanding that we would have used the bindings to provide a proper documentation to all the devices (not only sensors) for which a driver is already existing. 

Now, I do not have anything against this PR. I like it. But I think we would need to come up to a solution to what my original request was. So, if bindings are going to be deleted what we are going to use instead? Maybe provide scripts that look for "compaible" strings and provide a list of drivers? I think we should come up with a solution in #30104 if this PR will be accepted (which I think should).

So, I'm not approving it yet, but just to give a chance to at least have a new proposal for documenting the device supported.

**Comment by pabigot** [comment]
> First of all let me explain to you what is what I'm asking for at the first place. 

@avisconti Thanks, that helps.  I proposed this to address #31253, but that's a low priority bug.  Your needs are different, and I'm fine with having other infrastructure make this more visible and delaying this until there's a plan in place.  I just don't want to see bindings that do nothing but include the same properties as another device when both are supported by the same driver and we can use the standard compatible property ordering to describe what's going on.

**Comment by avisconti** [comment]
> > First of all let me explain to you what is what I'm asking for at the first place.
> 
> @avisconti Thanks, that helps. I proposed this to address #31253, but that's a low priority bug. Your needs are different, and I'm fine with having other infrastructure make this more visible and delaying this until there's a plan in place. I just don't want to see bindings that do nothing but include the same properties as another device when both are supported by the same driver and we can use the standard compatible property ordering to describe what's going on.

Sure! Let's see what @mbolivar-nordic and @erwango have to say about it. 
Marti, can we go in this direction where we have a script that extracts the list of all supported devices from the compatible string in bindings to create a proper documentation?

**Review by erwango — COMMENTED** [review]
As @avisconti explained, we should be able to document what hardware pieces are supported despite code factorization. And I still consider that compatibles should be the base to document that.

I'm fine with this PR as long as https://github.com/zephyrproject-rtos/zephyr/pull/30104 (or the next iteration) is able to take it into account and provide an adequate documentation out of it.

File: dts/bindings/sensor/st,lis2dh-i2c.yaml [hunk 2]
```diff
@@ -2,7 +2,30 @@
...
+
+    The LIS3DH sensor is supported by the same driver, and should be
+    specified in devicetree with:
+
+        compatible = "st,lis3dh", "st,lis2dh";
```
[the remark this answers is not in the document]
Comment (avisconti) [reply dts/bindings/sensor/st,lis2dh-i2c.yaml:12]: Like:

  compatible-list = "st,lsm303agr-accel", "st,lsm303dlhc-accel", "st,lis2dh12", "st,lis3dh", "st,lis2dh";

Something like that @erwango?

[the remark this answers is not in the document]
Comment (pabigot) [reply dts/bindings/sensor/st,lis2dh-i2c.yaml:12]: Also maybe:

    zephyr,driver-kconfig = "LIS2DH";

to complete the coupling between compatible and (default) driver.

[the remark this answers is not in the document]
Comment (erwango) [reply dts/bindings/sensor/st,lis2dh-i2c.yaml:12]: > compatible-list = "st,lsm303agr-accel", "st,lsm303dlhc-accel", "st,lis2dh12", "st,lis3dh", "st,lis2dh";

For instance. and that would be part of binding meta data.

> zephyr,driver-kconfig = "LIS2DH";

Still meta-data rigth ?

[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply dts/bindings/sensor/st,lis2dh-i2c.yaml:12]: >  new field in this compatible that provide the list of compatible compatibles

Wow, that's quite meta ;)

> compatible-list = "st,lsm303agr-accel", "st,lsm303dlhc-accel", "st,lis2dh12", "st,lis3dh", "st,lis2dh";

Where does this go; in the YAML? Not in DT, right?

**Comment by mbolivar-nordic** [comment]
> Marti, can we go in this direction where we have a script that extracts the list of all supported devices from the compatible string in bindings to create a proper documentation?

The main problem I was trying to solve in #30104 is the way to associate a driver implementation with its Kconfig options and DTS compatibles.

I tried to do this in Doxygen, but ran into various issues that make that approach impractical in my opinion.

So what I would like to do instead is to basically take the Doxygen comments in this commit:
https://github.com/zephyrproject-rtos/zephyr/pull/30104/commits/383c3ea2cb1359b3b62f7c2eb069179934db032d

And move them into .rst files in the drivers/ directory, the same way we have sample-specific documentation in the samples/ directory.

That will let us write free-form text to document our drivers that lets us bind together the Kconfig and DT compatibles. We can put that somewhere in the reference pages.

File: dts/bindings/sensor/st,lis2dh-i2c.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pabigot) [reply dts/bindings/sensor/st,lis2dh-i2c.yaml:12]: Yes, all this is intended to be YAML; I probably got the key/value syntax wrong.

**Comment by pabigot** [comment]
To summarize my high-level "requirements":

Given a sensor like LIS2DH12 I want a way to way to quickly identify:
* the compatible I need to identify the specific sensor.  I would expect to be able to find this by `git grep -i lis2dh12` in the tree, though it could also be in an index in generated documentation.  Here I would be looking for `st,lis2dh12`, so that text should show up.
* the compatible I need to specify the generic sensor that's used by the driver to provide the functionality.  This would be `st,lis2dh` in the example.  Seeing the recommended devicetree `compatible` property as in my proposal here would be ideal.
* the driver I need for the sensor, specifically the `KCONFIG` symbol necessary to enable it.

I'd expect to find all of that in one of the places revealed by the first search, either the generic sensor YAML, the specific sensor YAML if the specific sensor has additional properties, or the driver implementation file.

> So what I would like to do instead is to basically take the Doxygen comments in 383c3ea nd move them into .rst files in the drivers/ directory, the same way we have sample-specific documentation in the samples/ directory.

That'd work too.

File: dts/bindings/sensor/st,lis2dh-i2c.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (avisconti) [reply dts/bindings/sensor/st,lis2dh-i2c.yaml:12]: Yes, in YAML.

**Comment by avisconti** [comment]
> So what I would like to do instead is to basically take the Doxygen comments in this commit:
> [383c3ea](https://github.com/zephyrproject-rtos/zephyr/commit/383c3ea2cb1359b3b62f7c2eb069179934db032d)
> 
> And move them into .rst files in the drivers/ directory, the same way we have sample-specific documentation in the samples/ directory.
> 
> That will let us write free-form text to document our drivers that lets us bind together the Kconfig and DT compatibles. We can put that somewhere in the reference pages.

OK, let me try to see if I get your proposal. Bascally we need to provide a .rst file inside each driver folder (like driver/sensors/lis2dh) in which we are free to document whatever we feel is important to document. THis may be done with the help pof doxygen (even if I feel it is not 100% needed). But at the end in the zephyrproject documentation pages we will have a "driver supported" page beside to "Sample and Demos" and "Supported Boards". Is that what you mean?

**Comment by avisconti** [comment]
> Given a sensor like LIS2DH12 I want a way to quickly identify:
> 
> * the compatible I need to identify the specific sensor.  I would expect to be able to find this by `git grep -i lis2dh12` in the tree, though it could also be in an index in generated documentation.  Here I would be looking for `st,lis2dh12`, so that text should show up.
> * the compatible I need to specify the generic sensor that's used by the driver to provide the functionality.  This would be `st,lis2dh` in the example.  Seeing the recommended devicetree `compatible` property as in my proposal here would be ideal.
> * the driver I need for the sensor, specifically the `KCONFIG` symbol necessary to enable it.

YEs, and this is commented in the proper way in this commit itself provided by you. I mean, inside the lis2dh yaml file. Moreover, if I'm getting properly what @mbolivar-nordic is proposing, we will also have a .rst file in lis2dh driver which will describe all the compatibles required for each sensor supported by lis2dh driver.

**Comment by erwango** [comment]
> the compatible I need to identify the specific sensor. I would expect to be able to find this by git grep -i lis2dh12 in the tree, though it could also be in an index in generated documentation. Here I would be looking for st,lis2dh12, so that text should show up.

Let's target the generated documentation option. Based on my short community support experience, grepping the tree doesn't answer most of the questions we usually get (or if it would be an option we'd get much less questions).

> That will let us write free-form text to document our drivers that lets us bind together the Kconfig and DT compatibles. We can put that somewhere in the reference pages.

And these would be sorted by compatible in the documentation ?

**Comment by pabigot** [comment]
> we will also have a .rst file in lis2dh driver

If the free-form comments I put in the YAML of the base driver compatible are instead in RST, they don't have to be replicated in the YAML.  In fact, my proposal for fixing this PR would be to move them to such a file.

If we go that way, I don't think we need the `zephyr,driver-kconfig` YAML key.  I'm still thinking there *might* be a diagnostic value in having `compatible-list` in YAML, but we can wait until that's proven to be true.

**Comment by avisconti** [comment]
> > we will also have a .rst file in lis2dh driver
> 
> If the free-form comments I put in the YAML of the base driver compatible are instead in RST, they don't have to be replicated in the YAML. In fact, my proposal for fixing this PR would be to move them to such a file.

Yes, I think that having them in a single place reduce the future info maintenance.

> 
> If we go that way, I don't think we need the `zephyr,driver-kconfig` YAML key. I'm still thinking there _might_ be a diagnostic value in having `compatible-list` in YAML, but we can wait until that's proven to be true.
> 
> Edit: Actually, if we do have the driver Kconfig, we might be able to link from the generated YAML documentation to the corresponding generated driver documentation.

compatible-list was proposed to just have a way to automatically generate/document a list of devices supported by a given driver. I think we do not need it anymore if we go in the pre-driver .rst direction.

Concerning the other attribute for KConfig I'm not sure. Let's think about it.

Anyway, I have the feeling that we are not all 100% aligned. Am I wrong?
@erwango @mbolivar-nordic do you agree on all the raised points?

**Comment by erwango** [comment]
> compatible-list was proposed to just have a way to automatically generate/document a list of devices supported by a given driver. I think we do not need it anymore if we go in the per-driver .rst direction.

Which means we'd need .rst file per specific sensor, even of support is provided through generic driver ?
This could be fine for sensors as we generally have one driver per directory, for other drivers and subsystems, I'm affraid this would be much less clearer (several drivers per directory). But maybe I don't get the whole plan.

**Comment by pabigot** [comment]
> Which means we'd need .rst file per specific sensor, even of support is provided through generic driver ?

No, I assumed one `.rst` per driver/Kconfig selecting variable.  But you raise a good point; perhaps the name of the RST file for a driver should be the KConfig symbol that selects it, rather than `README.rst` in the directory.  So `LIS2DH.rst` would document all compatibles supported when `CONFIG_LIS2DH=y`.

**Comment by erwango** [comment]
>  So LIS2DH.rst would document all compatibles supported when CONFIG_LIS2DH=y.

Then for I2C, I assume we'd go for I2C.rst ( as opposed to I2C_STM32.rst, I2C_NRF.rst, ...).

This means in the general doc, we'd have 2 sets entries:
- Kconfig symbols, pointing to all available matching compatibles (while non-selectable DRIVER_VENDOR symbols would remain undocumented).
- Existing, per driver, generated compatibles entries.

This would solve the issue of having bindings  that do not content Zephyr specific Kconfig symbols.
Would it be compatible with a possible automated dts based board documentation ?

**Comment by pabigot** [comment]
> > So LIS2DH.rst would document all compatibles supported when CONFIG_LIS2DH=y.
> 
> Then for I2C, I assume we'd go for I2C.rst ( as opposed to I2C_STM32.rst, I2C_NRF.rst, ...).

? No, I was proposing using the driver Kconfig, not the subsystem Kconfig, so probably `I2C_STM32_V1.rst`, `I2C_STM32_V2.rst`.  But that breaks down for Nordic, which has `I2C_NRFX` but selects between the `TWI` and `TWIM` drivers on a per-instance basis based on compatible.

Ugh.  Something to revisit in 2.6.

**Comment by erwango** [comment]
> No, I was proposing using the driver Kconfig, not the subsystem Kconfig, so probably I2C_STM32_V1.rst, I2C_STM32_V2.rst. But that breaks down for Nordic, which has I2C_NRFX but selects between the TWI and TWIM drivers on a per-instance basis based on compatible.

I'm not quite for this proposal. To me, these symbols shouldn't be user selectable, user should only care to enable I2C and the matching I2C_VENDOR should be selected, either through Kconfig dependencies, either thanks to identification of enabled i2c compatibles in board dts.
And since user shouldn't care, I don't think these should be a documentation entry point.

**Comment by pabigot** [comment]
I do think a user should always be able to tell, given a compatible, what driver supports it, and vice-versa.

**Comment by avisconti** [comment]
> I do think a user should always be able to tell, given a compatible, what driver supports it, and vice-versa.

Yes, I agree. In linux the compatible string is in the driver itself. In zephyr case is in YAML.
But I do believe it is a limitation. So, I think we definitely need to have .rst per driver which provide the list of all compatibles it supports.

So, I guess that going with something like lis2dh.rst, i2c_stm32_v1.rst, ... would be appropriate.
Ideally I would see a triplet of driver, binding, doc, e.g. <st,lis2dh.yaml, CONFIG_LIS2DH, lis2dh.rst>.
IN case of I2C <st,stm32-i2c-v1.yaml, CONFIG_I2C_STM32_V1, i2c_ll_stm32_v1.rst>. Does it make sense to you?

**Comment by erwango** [comment]
> IN case of I2C <st,stm32-i2c-v1.yaml, CONFIG_I2C_STM32_V1, i2c_ll_stm32_v1.rst>. Does it make sense to you?

I'm affraid this would be misleading.
One of Zephyr highlights is the application portability. So application shouldn't care if I2C driver in use is I2C_STM32_V1, I2C_STM32_V2 or I2C_NRF, it should only car to enable I2C=y is I2C is required.
Thanks to device tree and Kconfig magic, we're able to automatically enable the right I2C_VENDOR symbol that matches the targeted device. This is how things works in > 95% of current tests and samples.

Documenting things as proposed: <st,stm32-i2c-v1.yaml, CONFIG_I2C_STM32_V1, i2c_ll_stm32_v1.rst> puts into light vendor symbol that user shouldn't care is majority of cases and omit the important portability information ("driver requires I2C=y").
So, if really needed vendor symbols could be mentioned, but it should be clear that the main symbol to be used is "I2C".

**Comment by avisconti** [comment]
> > IN case of I2C <st,stm32-i2c-v1.yaml, CONFIG_I2C_STM32_V1, i2c_ll_stm32_v1.rst>. Does it make sense to you?
> 
> I'm affraid this would be misleading.
> One of Zephyr highlights is the application portability. So application shouldn't care if I2C driver in use is I2C_STM32_V1, I2C_STM32_V2 or I2C_NRF, it should only car to enable I2C=y is I2C is required.
> Thanks to device tree and Kconfig magic, we're able to automatically enable the right I2C_VENDOR symbol that matches the targeted device. This is how things works in > 95% of current tests and samples.
> 
> Documenting things as proposed: <st,stm32-i2c-v1.yaml, CONFIG_I2C_STM32_V1, i2c_ll_stm32_v1.rst> puts into light vendor symbol that user shouldn't care is majority of cases and omit the important portability information ("driver requires I2C=y").
> So, if really needed vendor symbols could be mentioned, but it should be clear that the main symbol to be used is "I2C".

I think that we are talking about different requirements. Usually one may want to know if a give sensor is supported, but usually does not care whether a particular I2C controller is supported or not. In that case usually one may just want to know if a particular SoC or even board is supported. So, in this case would a single I2C.rst file that is documenting all compatibles for all I2C bus controllers better? 

What I feel is really necessary is to have a single page that lists all devices supported and what is the driver associated. This should be easily extracted from the compatibles. How to do that I don't know exactly. Moreover the idea of having a rst for each driver does not solve the immediate doubt whether lis2dh12 is supported or not, because you have to know in advance that is lis2dh.

So, I'm coming back with the original idea to automatically extract from the bindings the list of compatibles and create such a page.

**Comment by mbolivar-nordic** [comment]
@avisconti 

>  Bascally we need to provide a .rst file inside each driver folder (like driver/sensors/lis2dh) in which we are free to document whatever we feel is important to document.

Yes.

> THis may be done with the help pof doxygen (even if I feel it is not 100% needed). 

No. Maybe the role of doxygen is confusing because #30104 used it?

TL;DR I tried to use doxygen in #30104 but it didn't/doesn't work. So I **do not** think we should use doxygen for this. See #30104 for details.

> But at the end in the zephyrproject documentation pages we will have a "driver supported" page

Yes, but...

> beside to "Sample and Demos" and "Supported Boards". Is that what you mean?

No, I'm not proposing that location; what I wrote was:

> We can put that somewhere in the reference pages.

And by that I meant here: https://docs.zephyrproject.org/latest/reference/index.html. This is where the APIs are documented.

Thinking about it more I believe we should have a list of drivers that we link to from the documentation for each API. So for example https://docs.zephyrproject.org/latest/reference/peripherals/adc.html#adc-api would link to the drivers implementing the API, or perhaps link to an index page of all the drivers.

**Comment by avisconti** [comment]
> @avisconti
> 
> > Bascally we need to provide a .rst file inside each driver folder (like driver/sensors/lis2dh) in which we are free to document whatever we feel is important to document.
> 
> Yes.
> 
> 
> Thinking about it more I believe we should have a list of drivers that we link to from the documentation for each API. So for example https://docs.zephyrproject.org/latest/reference/peripherals/adc.html#adc-api would link to the drivers implementing the API, or perhaps link to an index page of all the drivers.

@mbolivar-nordic 
Documenting some of the driver, for example lis2dh, is only a part of the original issue I was reporting. In this way a user may find what lis2dh (for example) may be used for.

But there is a second part which is currently not covered yet, and that is how a user may find the information if a given device is currently supported or not, and by which driver. And the example I had in mind was lis2dh12, which is supported by lis2dh. If we only document lis2dh driver the user has to guess that lis2dh driver is covering more than one device. Instead, I think we should have also a page where all supported devices are listed. At the end is a compatible list.

**Comment by mbolivar-nordic** [comment]
> Instead, I think we should have also a page where all supported devices are listed. At the end is a compatible list.

I imagined that a driver's documentation would include a list of compatibles it supported, so they could be found via freetext search on the docs site. But I have no objections to a generated index of the same; I think Sphinx has the support we need for auto-generating indices.

**Comment by avisconti** [comment]
> > Instead, I think we should have also a page where all supported devices are listed. At the end is a compatible list.
> 
> I imagined that a driver's documentation would include a list of compatibles it supported, so they could be found via freetext search on the docs site. But I have no objections to a generated index of the same; I think Sphinx has the support we need for auto-generating indices.

That would be great. I'm not sure how many people would grep the device name in the source code to see if some driver is supporting it.

**Comment by pabigot** [comment]
> That would be great. I'm not sure how many people would grep the device name in the source code to see if some driver is supporting it.

It was one of [my requirements](https://github.com/zephyrproject-rtos/zephyr/pull/31450#issuecomment-765954821).

**Comment by pabigot** [comment]
I'm not working to solve this anymore and this is far too old to merge.
